File: interview_java/interview/listarray/add2number.java
package listarray;

import java.util.*;

public class add2number {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    	if (l1 == null && l2 == null) return null;
    	if (l1 == null && l2 != null) return l2;
    	if (l2 == null && l1 != null) return l1;
    	
        int sum, digit, carry;
        ListNode p1, p2, curr;
        ListNode head = new ListNode(0);
        curr = head;
        p1 = l1; p2 = l2;
        digit = 0;
        carry = 0;
        
        while(p1 != null && p2 != null) {
        	sum = p1.val+p2.val + carry;
        	digit = sum%10;
        	carry = sum/10;
        	ListNode tmp = new ListNode(digit);
        	curr.next = tmp;
        	p1 = p1.next;
        	p2 = p2.next;
        	curr = curr.next;
        }        
        
        if (p1 != null) curr.next = p1;
        else if (p2 != null) curr.next = p2;
                
    	return head.next;
    }
    
    public static void main(String[] args){
    	
    	System.out.println("Java:add2sum");
    	
    	add2number test = new add2number();
    	
    	ListNode l1 = new ListNode(2);
    	l1.next = new ListNode(4);
    	l1.next.next = new ListNode(3);
    	
    	ListNode l2 = new ListNode(5);
    	l2.next = new ListNode(6);
    	l2.next.next = new ListNode(4);
    	
    	ListNode result = test.addTwoNumbers(l1, l2);
    	
    	while (result != null) {
    		System.out.println(result.val);
    		result = result.next;
    	}
    	
    }
    
}

File: interviewcpp/others/add2number/add2number.cpp
/*
 * add2number.cpp
 *
 *  Created on: Oct 26, 2016
 *      Author: songjiguo
 */

using namespace std;

#include <iostream>

struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(nullptr) {}
};

class add2number {
public:
	add2number();
	virtual ~add2number();

	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2);
};


ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
	 if (!l1 && !l2) return nullptr;
	 if (l1 && !l2) return l1;
	 if (l2 && !l1) return l2;

	 int sum, digit, carry;
	 carry = 0;
	 ListNode *head = nullptr;
	 ListNode *curr = head;

	 ListNode *p1, *p2;
	 p1 = l1;
	 p2 = l2;

	 while (p1 && p2) {
		 sum = p1->val + p2->val;
		 digit = sum%10 + carry;
		 carry = sum/10;
		 ListNode *tmp = new ListNode(digit);
		 if (!head) {
			 head = tmp;
			 curr = tmp;
		 }  else {
			 curr->next = tmp;
			 curr = curr->next;
		 }
		 p1 = p1->next;
		 p2 = p2->next;
	 }

	 if (p1) curr->next = p1;
	 if (p2) curr->next = p2;

	 return head;
}

int main() {

	cout << "this is another test";
	cout << "CPP:add2nums" << endl;
	ListNode *l1 = new ListNode(2);
	l1->next = new ListNode(4);
	l1->next->next = new ListNode(3);

	ListNode *l2 = new ListNode(5);
	l2->next = new ListNode(6);
	l2->next->next = new ListNode(4);

	ListNode *tmp = addTwoNumbers(l1, l2);
	while(tmp) {
		cout << tmp->val << endl;
		tmp = tmp->next;
	}

	return 0;
}

